internal/resolver/dns/fake_net_resolver_test.go

```diff
@@ -0,0 +1,579 @@
+/*
+ *
+ * Copyright 2023 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package dns_test
+
+import (
+	"context"
+	"net"
+	"sync"
+
+	"google.golang.org/grpc/internal/testutils"
+)
+
+// A fake implementation of the internal.NetResolver interface for use in tests.
+type testNetResolver struct {
+	// A write to this channel is made when this resolver receives a resolution
+	// request. Tests can rely on reading from this channel to be notified about
+	// resolution requests instead of sleeping for a predefined period of time.
+	lookupHostCh *testutils.Channel
+}
+
+func (tr *testNetResolver) LookupHost(ctx context.Context, host string) ([]string, error) {
+	if tr.lookupHostCh != nil {
+		tr.lookupHostCh.Send(nil)
+	}
+	return hostLookup(host)
+}
+
+func (*testNetResolver) LookupSRV(ctx context.Context, service, proto, name string) (string, []*net.SRV, error) {
+	return srvLookup(service, proto, name)
+}
+
+func (*testNetResolver) LookupTXT(ctx context.Context, host string) ([]string, error) {
+	return txtLookup(host)
+}
+
+var hostLookupTbl = struct {
+	sync.Mutex
+	tbl map[string][]string
+}{
+	tbl: map[string][]string{
+		"foo.bar.com":          {"1.2.3.4", "5.6.7.8"},
+		"ipv4.single.fake":     {"1.2.3.4"},
+		"srv.ipv4.single.fake": {"2.4.6.8"},
+		"srv.ipv4.multi.fake":  {},
+		"srv.ipv6.single.fake": {},
+		"srv.ipv6.multi.fake":  {},
+		"ipv4.multi.fake":      {"1.2.3.4", "5.6.7.8", "9.10.11.12"},
+		"ipv6.single.fake":     {"2607:f8b0:400a:801::1001"},
+		"ipv6.multi.fake":      {"2607:f8b0:400a:801::1001", "2607:f8b0:400a:801::1002", "2607:f8b0:400a:801::1003"},
```

Comment:
Hmm, I was hoping this and the below table would go away in any rewrite.  Is there any way to make this declared in a constructor instead?  If it's too much work we can punt, but IMO splitting test code that depends on each other into very distant places is a very bad practice.

Reply:
I will look into this. But if we can do that as part of a follow up PR, that would be good, since many of the other xDS PRs are gated on the change to add `testutils.ResolverClientConn`.

Reply:
Improved the host lookup and srv lookup parts. Working on improving the txt lookup.